notebooks/knn.ipynb

```diff
@@ -0,0 +1,399 @@
+    "from pyspark.sql.functions import col\n",
+    "\n",
+    "item_vector_df = item_df.select(array_to_vector(item_df.features).alias(\"features\"))\n",
+    "query_vector_df = query_df.select(array_to_vector(query_df.features).alias(\"features\"))\n",
+    "key = Vectors.dense([1.0, 1.0])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f4288d38-6426-41c0-87a2-a75bc9d4bbda",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "item_vector_df.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "655f8071-c90e-464b-aa25-58d96bbeebea",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "query_vector_df.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c8712f3f-4662-409e-8172-29308ec84e0c",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "brp = BucketedRandomProjectionLSH(inputCol=\"features\", outputCol=\"hashes\", bucketLength=2.0, numHashTables=3)\n",
+    "model = brp.fit(item_vector_df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4aa5813b-41b9-4b51-9977-721e6f90118e",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Feature Transformation\n",
+    "print(\"The hashed dataset where hashed values are stored in the column 'hashes':\")\n",
+    "model.transform(item_vector_df).show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "90f7049a-5ed1-44a8-b0d0-be63f6458e0a",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Compute the locality sensitive hashes for the input rows, then perform approximate similarity join.\n",
+    "# We could avoid computing hashes by passing in the already-transformed dataset, e.g.\n",
+    "# `model.approxSimilarityJoin(transformedA, transformedB, 1.5)`\n",
+    "print(\"Approximately joining items and queries on Euclidean distance smaller than 3.0:\")\n",
+    "model.approxSimilarityJoin(item_vector_df, query_vector_df, 3.0, distCol=\"EuclideanDistance\")\\\n",
+    "    .select(col(\"datasetA.features\").alias(\"item\"),\n",
+    "            col(\"datasetB.features\").alias(\"query\"),\n",
+    "            col(\"EuclideanDistance\")).orderBy(\"query\", \"item\").show()\n",
+    "\n",
+    "# results = model.approxSimilarityJoin(item_vector_df, query_vector_df, 1.5, distCol=\"EuclideanDistance\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e70d24b7-db70-4041-9794-7cd5451bad76",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Compute the locality sensitive hashes for the input rows, then perform approximate nearest neighbor search.\n",
+    "# We could avoid computing hashes by passing in the already-transformed dataset, e.g.\n",
+    "# `model.approxNearestNeighbors(transformedA, key, 2)`\n",
```

Comment:
was there a transformedA earlier?

Reply:
transformedA seems a term used to comment CPU example codes: https://spark.apache.org/docs/2.2.3/ml-features.html#lsh-operations.  It is not a variable.

Reply:
Fixed to reflect variables in notebook.